server/server/src/billing/billing_service.rs

```diff
@@ -381,113 +504,259 @@ pub async fn stripe_webhooks(
+        .maybe_json_get(public_key_from_stripe_customer_id(&customer_id))
         .await?
         .ok_or_else(|| {
             internal!(
-                "Payment failed for a customer we don't have info about on redis: {:?}",
-                event
+                "There is no public_key related to this customer_id: {:?}",
+                customer_id
             )
         })?;
 
-    Ok((public_key, user_info))
+    Ok(public_key)
+}
+
+#[derive(Debug)]
+pub enum GooglePlayWebhookError {
+    InvalidToken,
+    CannotRetrieveData,
+    CannotDecodePubSubData(DecodeError),
+    CannotRetrieveUserInfo,
+    CannotRetrievePublicKey,
+    CannotParseTime,
+}
+
+pub async fn android_notification_webhooks(
+    server_state: &Arc<ServerState>, request_body: Bytes, auth_token: String,
+) -> Result<(), ServerError<GooglePlayWebhookError>> {
+    if !constant_time_eq::constant_time_eq(
+        auth_token.as_bytes(),
+        server_state.config.google.pubsub_token.as_bytes(),
+    ) {
+        return Err(ClientError(GooglePlayWebhookError::InvalidToken));
+    }
+
+    let message = serde_json::from_slice::<PubsubMessage>(&request_body)?;
+    let data = base64::decode(
+        message
+            .data
+    )
+    .map_err(|e| ClientError(GooglePlayWebhookError::CannotDecodePubSubData(e)))?;
+
+    let notification = serde_json::from_slice::<DeveloperNotification>(&data)?;
+
+    let mut con = server_state.index_db_pool.get().await?;
+
+    if let Some(sub_notif) = notification.subscription_notification {
+        let purchase = google_play_client::get_subscription(
+            &server_state.android_publisher,
+            &sub_notif.subscription_id,
+            &sub_notif.purchase_token,
+        )
+        .await
+        .map_err(|e| match e {
+            SimpleGCPError::Unexpected(msg) => internal!("{:#?}", msg),
+        })?;
+
+        let public_key: PublicKey =
+            serde_json::from_str(&purchase.developer_payload.clone().ok_or(internal!(
+                "There should be a public key attached to a purchase: {:?}",
+                sub_notif.subscription_id
+            ))?)?;
+
+        let notification_type = sub_notif.notification_type();
+        let mut billing_hist_size: Option<usize> = None;
+
+        let tx_result = tx!(&mut con, pipe, &[keys::billing_lock(&public_key)], {
+            match lock_billing_workflow(
+                &public_key,
+                &mut con,
+                server_state.config.stripe.millis_between_user_payment_flows,
+            )
+            .await {
+                Ok(mut billing_lock) => {
+                    if let Some(size) = billing_hist_size {
+                        if size != billing_lock.info.len() {
+                            return Ok(&mut pipe);
+                        }
+                    }
+
+                    billing_hist_size = Some(billing_lock.info.len());
+                    if let Some(BillingInfo::Stripe(info)) = billing_lock.info.last_mut() {
```

Comment:
looks like a fatal copy paste mistake to me?

Reply:
What is the mistake?

Reply:
`android_notification_webhooks()`

`BillingInfo::Stripe(info)`

Reply:
Oh, that is definitely wrong. Thanks!